substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfLocSectionImpl.java

```diff
@@ -326,6 +334,56 @@ private int writeStackLocation(DebugContext context, int offset, byte[] buffer,
         return pos;
     }
 
+    private int writeConstantLocation(DebugContext context, JavaConstant constant, byte[] buffer, int p) {
+        int pos = p;
+        byte op = DW_OP_implicit_value;
+        JavaKind kind = constant.getJavaKind();
+        int byteCount = (kind == JavaKind.Object ? 8 : kind.getByteCount());
+        assert (kind != JavaKind.Object || constant.isNull()) : "only expecting null object constant!";
+        if (buffer == null) {
+            pos += putByte(op, scratch, 0);
+            pos += putULEB(byteCount, scratch, 0);
+            if (constant.isNull()) {
+                pos += writeAttrData8(0, scratch, 0);
+            } else if (byteCount == 1) {
+                if (kind == JavaKind.Int) {
+                    pos += putByte((byte) constant.asInt(), scratch, 0);
+                } else {
+                    pos += putByte((byte) (constant.asBoolean() ? 1 : 0) , scratch, 0);
+                }
+            } else if (byteCount == 2) {
+                pos += putShort((short) constant.asInt(), scratch, 0);
+            } else if (byteCount == 4) {
+                int i = (kind == JavaKind.Int ? constant.asInt() : Float.floatToRawIntBits(constant.asFloat()));
+                pos += putInt(i, scratch, 0);
+            } else {
+                long l =  (kind == JavaKind.Long ? constant.asLong() : Double.doubleToRawLongBits(constant.asDouble()));
+                pos += putLong(l, scratch, 0);
+            }
+        } else {
```

Comment:
If we would patch all the `put*` methods in `com.oracle.objectfile.elf.dwarf.DwarfSectionImpl` with something like
```patch
--- a/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfSectionImpl.java	(revision Staged)
+++ b/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfSectionImpl.java	(date 1651057782970)
@@ -147,10 +147,12 @@
      * Base level put methods that assume a non-null buffer.
      */
 
-    protected int putByte(byte b, byte[] buffer, int p) {
-        int pos = p;
+    protected int putByte(byte b, byte[] buf, int p) {
+        boolean useScratch = buf == null;
+        int pos = useScratch ? 0 : p;
+        byte[] buffer = useScratch ? scratch : buf;
         buffer[pos++] = b;
-        return pos;
+        return pos + (useScratch ? p : 0);
     }
 
     protected int putShort(short s, byte[] buffer, int p) {
```
we could get collapse these almost identical cases for `buffer == null` and `buffer != null`

Reply:
> Yes, I agree that would be a very good idea, although I would prefer to do it in a follow-up PR.

Fine with me. I created https://github.com/oracle/graal/issues/4513 for that.